src/js/components/AppListItemLabelsComponent.jsx

```diff
@@ -0,0 +1,159 @@
+var classNames = require("classnames");
+var React = require("react/addons");
+
+var OnClickOutsideMixin = require("react-onclickoutside");
+var Util = require("../helpers/Util");
+
+var AppListItemLabelsComponent = React.createClass({
+  displayName: "AppListItemLabelsComponent",
+
+  initialMarginTop: null,
+
+  mixins: [OnClickOutsideMixin],
+
+  propTypes: {
+    children: React.PropTypes.node,
+    labels: React.PropTypes.object,
+    numberOfVisibleLabels: React.PropTypes.number
+  },
+
+  getInitialState: function () {
+    return {
+      isDropdownVisible: false
+    };
+  },
+
+  componentDidUpdate: function () {
+    if (this.state.isDropdownVisible === true) {
+      this.recalculateDropdownPosition();
+    }
+  },
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return this.didPropsChange(nextProps, nextState);
+  },
+
+  didPropsChange: function (props, state) {
+    let keys = ["labels", "numberOfVisibleLabels"];
+    return !Util.compareProperties(this.props, props, ...keys) ||
+      this.state.isDropdownVisible !== state.isDropdownVisible;
+  },
+
+  handleClickOutside: function () {
+    this.setState({
+      isDropdownVisible: false
+    });
+  },
+
+  handleShowMoreClick: function (event) {
+    event.stopPropagation();
+    this.setState({
+      isDropdownVisible: !this.state.isDropdownVisible
+    });
+  },
+
+  recalculateDropdownPosition: function () {
+    if (global.window == null) {
+      return;
+    }
+
+    let dropdownNode = React.findDOMNode(this.refs.labelsDropdown);
+    let leftArrowNode = React.findDOMNode(this.refs.leftArrow);
+
+    // First time the dropdown becomes visible: get its initial top offset
+    if (this.initialMarginTop == null) {
```

Comment:
Couldn't this be a local var? Would eliminate you class-member problem.
I think it's only slightly slower.